ledger/participant-integration-api/src/test/suite/scala/platform/store/completions/PagedCompletionsReaderWithCacheSpec.scala

```diff
@@ -0,0 +1,315 @@
+    val pastCompletions = await(
+      reader.getCompletionsPage(
+        genOffset(9),
+        genOffset(19),
+        ApplicationId.assertFromString("Application_1_Id"),
+        Set(Party.assertFromString("party1")),
+      )
+    )
+    // THEN: correct number of completions should be returned
+    pastCompletions.size shouldBe 10
+
+    // AND database should be queried with application and parties filter
+    completionsDao.runQueriesJournalToSeq() shouldBe Seq[Request](
+      GetAllCompletionsRequest(genOffset(9), genOffset(19)),
+      GetAllCompletionsRequest(genOffset(19), genOffset(30)),
+      GetFilteredCompletionsRequest(
+        genOffset(9),
+        genOffset(19),
+        ApplicationId.assertFromString("Application_1_Id"),
+        Set(Party.assertFromString("party1")),
+      ),
+    )
+  }
+
+  it should "correctly fetch data overlapping with cache boundaries" in {
+
+    // GIVEN: completions reader and dao mock with 40 completions
+    val completionsDao = new CompletionsDaoMock(CompletionsDaoMock.genCompletions(40))
+    val reader = new PagedCompletionsReaderWithCache(completionsDao, 10)
+
+    // WHEN: not cached completions are fetched
+    val cachedCompletions = await(
+      reader.getCompletionsPage(
+        genOffset(19),
+        genOffset(29),
+        ApplicationId.assertFromString("Application_1_Id"),
+        Set(Party.assertFromString("party1")),
+      )
+    )
+
+    // THEN: correct number of completions should be returned
+    cachedCompletions.size shouldBe 10
+
+    // WHEN completions for range bigger than cache size are requested
+    val overlappingCompletions = await(
+      reader.getCompletionsPage(
+        genOffset(9),
+        genOffset(39),
+        ApplicationId.assertFromString("Application_1_Id"),
+        Set(Party.assertFromString("party1")),
+      )
+    )
+
+    // THEN: correct number of completions should be returned
+    overlappingCompletions.size shouldBe 30
+
+    // AND database should be queried with application and parties filter for older than cached completions (<19)
+    // AND database should not be queried by cached completions (19, 29]
+    // AND database should be queried for all completions newer than 29
+    completionsDao.runQueriesJournalToSeq() shouldBe Seq[Request](
+      GetAllCompletionsRequest(genOffset(19), genOffset(29)),
+      GetFilteredCompletionsRequest(
+        genOffset(9),
+        genOffset(19),
+        ApplicationId.assertFromString("Application_1_Id"),
+        Set(Party.assertFromString("party1")),
+      ),
+      GetAllCompletionsRequest(genOffset(29), genOffset(39)),
+    )
+  }
+
+  it should "fetch each completion exactly once" in {
+    // GIVEN: completions reader and dao mock with 20 completions
+    val completionsDao = new CompletionsDaoMock(CompletionsDaoMock.genCompletions(20))
+    val reader = new PagedCompletionsReaderWithCache(completionsDao, 10)
+
+    // WHEN 10 equal requests are performed concurrently
```

Comment:
TODO: test concurrent requests that are not equal. For example, a request from (10, 12) followed by a request for (8,14). We want to test that the follow-up properly splits the data.

Reply:
It doesn't look like thing to unit test - if we have concurrent requests (10, 12) and (8,14) then cached results will be different depending on order. I would expect to be cached:

* (10, 14) for request order (10, 12),  (8,14)
* (8, 14) for request order (8,14), (10, 12)

It may work for request with the same start offset: (8, 12), (8, 14), but not sure if it brings value